### PR TITLE
tests: Add scheduled smoke tests for sentry.io

### PR DESCRIPTION
**Pull request by BYK**
Refs #27009

### CODE REVIEW

**Comment by EvanPurkhiser** [comment]
Hmm, should this be part of the sentry repository or should it live somewhere else? This actually looks more like it's smoke testing the marketing site to some degree

What happens when it fails, does it report somewhere to us?

**Comment by BYK** [comment]
@EvanPurkhiser the idea is to get this to run:
1. Regularly, every 5 minutes as you see here
2. On every PR with the Vercel preview thing if possible
3. After every deploy of getsentry

Happy to move it to a dedicated repo if you think that's more appropriate.

**Comment by BYK** [comment]
>What happens when it fails, does it report somewhere to us?

We get an email about it but I think we can use some of @armenzg's magic from hackweek to feed these into Sentry and set up alerts.

File: .github/workflows/smoke-tests.yml [hunk 1]
```diff
@@ -0,0 +1,25 @@
+name: 'Run Smoke Tests'
+on:
+  schedule:
+    - cron: '*/5 * * * *'
+  workflow_dispatch:
```
Comment (armenzg) [inline .github/workflows/smoke-tests.yml:5]: @EvanPurkhiser With one of our bots, we can hit an API that allows triggering this.

As @BYK mentioned, we can listen to these kind of events and decide what we want to do with it.

File: .github/workflows/smoke-tests.yml [hunk 2]
```diff
@@ -0,0 +1,25 @@
+name: 'Run Smoke Tests'
```
Comment (armenzg) [inline .github/workflows/smoke-tests.yml:1]: Neither the scheduled nor the workflow dispatch runs will show up on the PR as one of the associated checks.

We will need some tooling magic to notify folks in their PRs.

A run can be seen here (reachable by visiting the Actions tab):
https://github.com/getsentry/sentry/runs/3414995500?check_suite_focus=true

File: .github/workflows/smoke-tests.yml [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+jobs:
+  smoke_tests:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+      - uses: actions/setup-node@v2
+        with:
+          node-version: '14'
+      - name: Install operating system dependencies
+        run: npx playwright install-deps
```
Comment (armenzg) [inline .github/workflows/smoke-tests.yml:15]: We will need caching and pinning for the browser versions in order to speed up the set up time. We want smoke tests to run as fast as possible after a deployment (preferably within a minute). If we want to we could even have a slow run for known browser versions and a fast run for the pinned browser versions.

Perhaps we can have a Docker image with everything already set up in order to take off sooner.

image

image

Do the tests run against all 4 browsers?
Does it say anywhere against what version of each?

File: .github/workflows/smoke-tests.yml [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+      - uses: actions/cache@v2
+        id: cache
+        with:
+          path: node_modules
+          key: ${{ runner.os }}-${{ hashFiles('package.json', 'yarn.lock') }}
```
Comment (armenzg) [inline .github/workflows/smoke-tests.yml:20]: Do we want `api-docs/yarn.lock`?

File: .github/workflows/smoke-tests.yml [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+      - name: Install Dependencies
+        if: steps.cache.outputs.cache-hit != 'true'
+        run: yarn install --frozen-lockfile
```
Comment (armenzg) [inline .github/workflows/smoke-tests.yml:23]: It seems that this is necessary in order to allow executing the next step:
```
❯ npx playwright test --browser all tests/smoke
Please install @playwright/test package to use Playwright Test.
  npm install -D @playwright/test
```

Do we need to install all Sentry dependencies? Could we *just* install `@playwright/test`?

File: .github/workflows/smoke-tests.yml [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+      - name: Run tests
+        run: npx playwright test --browser all tests/smoke
```
Comment (armenzg) [inline .github/workflows/smoke-tests.yml:25]: It seems the framework executes the tests in parallel. Excellent!

```shell
sentry on  byk/ci/smoke-tests [$!] via  v12.19.0 via 🐍 v3.8.10 (.venv) on ☁️  armenzg@sentry.io took 24s
❯ time npx playwright test --browser all tests/smoke

Running 3 tests using 3 workers

  ✓  [chromium] › tests/smoke/sentryio.spec.ts:3:1 › sentryio (11s)
  ✓  [firefox] › tests/smoke/sentryio.spec.ts:3:1 › sentryio (13s)
  ✓  [webkit] › tests/smoke/sentryio.spec.ts:3:1 › sentryio (12s)


  3 passed (13s)
npx playwright test --browser all tests/smoke  16.14s user 2.61s system 96% cpu 19.463 total 
```

File: tests/smoke/sentryio.spec.ts [hunk 7]
```diff
@@ -0,0 +1,34 @@
+const {test, expect} = require('@playwright/test');
+
+test('sentryio', async ({page}) => {
+  // Go to https://sentry.io/
+  await page.goto('https://sentry.io/');
```
Comment (armenzg) [inline tests/smoke/sentryio.spec.ts:5]: I'm so impressed with how easy this framework makes this.

File: .github/workflows/smoke-tests.yml [hunk 8]
```diff
@@ -0,0 +1,25 @@
+name: 'Run Smoke Tests'
+on:
+  schedule:
+    - cron: '*/5 * * * *'
```
Comment (armenzg) [inline .github/workflows/smoke-tests.yml:4]: Currently, the way GH implements this, the last person to *touch* the workflow will get notified if this fails.

**Review by armenzg — APPROVED** [review]
Excellent work @BYK !
Most of my actionable comments are with regards to optimizing the set up time and how to integrate this into development workflow.

**Review by BYK — COMMENTED** [review]
Holding off for final feedback, and especially on the `docker` image approach and separate repo.

Now that I think about it, we can build our own docker image based of off the one I linked, deploy that with GCR or GCF and then trigger it using GCP scheduler. Would probably be a lot more efficient.

Thoughts?

File: .github/workflows/smoke-tests.yml (hunk 4 above)
Comment (billyvg) [inline .github/workflows/smoke-tests.yml:20]: @BYK FYI You can now enable caching from `action/setup-node` -> https://github.com/actions/setup-node/blob/main/docs/advanced-usage.md#caching-packages-dependencies

File: tests/smoke/sentryio.spec.ts [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+  // Click #sso >> text=Continue
+  await page.click('#sso >> text=Continue');
+  expect(page.url()).toBe('https://sentry.io/auth/login/sentry/');
+
+  // Click text=Login with Google
+  await page.click('text=Login with Google');
+  await page.waitForURL('https://accounts.google.com/**/*');
+
+  // Close page
+  await page.close();
```
Comment (billyvg) [inline tests/smoke/sentryio.spec.ts:33]: Is it a bad practice to have a test account in prod where our smoke tests can log into? Reason being, this auth page does not load our SPA.

**Review by billyvg — COMMENTED** [review]
good stuff :)

**Pull request closed without merging**